extension/observer/ecsobserver/task.go

```diff
@@ -65,3 +80,67 @@ func (t *Task) ContainerLabels(containerIndex int) map[string]string {
 	}
 	return labels
 }
+
+// PrivateIP returns private ip address based on network mode.
+// EC2 launch type can use hsot/brige mode and the private ip is the EC2 instance's ip.
+// awsvpc has its own ip regardless of launch type.
+func (t *Task) PrivateIP() (string, error) {
+	arn := aws.StringValue(t.Task.TaskArn)
+	switch aws.StringValue(t.Definition.NetworkMode) {
+	// Default network mode is bridge on EC2
+	case "", ecs.NetworkModeHost, ecs.NetworkModeBridge:
+		if t.EC2 == nil {
+			return "", fmt.Errorf("task has no network mode and no ec2 info %s", arn)
+		}
+		return aws.StringValue(t.EC2.PrivateIpAddress), nil
+	case ecs.NetworkModeAwsvpc:
+		for _, v := range t.Task.Attachments {
+			if aws.StringValue(v.Type) == "ElasticNetworkInterface" {
+				for _, d := range v.Details {
+					if aws.StringValue(d.Name) == "privateIPv4Address" {
+						return aws.StringValue(d.Value), nil
+					}
+				}
+			}
+		}
+		return "", fmt.Errorf("private ipv4 address not found for awsvpc on task %s", arn)
+	case ecs.NetworkModeNone:
+		return "", fmt.Errorf("task has none network mode %s", arn)
+	default:
+		return "", fmt.Errorf("unknown task network mode %q for task %s", aws.StringValue(t.Definition.NetworkMode), arn)
+	}
+}
+
+// MappedPort returns 'external' port based on network mode.
+// EC2 bridge gets random host port while EC2 host/awsvpc uses whatever specified by user.
+func (t *Task) MappedPort(def *ecs.ContainerDefinition, containerPort int64) (int64, error) {
+	arn := aws.StringValue(t.Task.TaskArn)
+	mode := aws.StringValue(t.Definition.NetworkMode)
+	switch mode {
+	case ecs.NetworkModeNone:
+		return 0, fmt.Errorf("task has none network mode %s", arn)
```

Comment:
For the errors, I plan to define error structs later to keep the PR smaller. They will change to dedicated error types for the telemetry https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/extension/observer/ecsobserver#metrics

Reply:
Changed to error struct. Might add an interface to indicates if this a fatal error. Though the error handling logic for the extension is basically log and ignore, because you can have invalid task/extension config while still getting some valid targets. Based on experience form existing customers (in cloudwatch agent), they don't want to stop the collection when some invalid task shows up (e.g. write a docker label `ECS_PROM_PORT: 123abc`).

The plan is to export those errors are metrics https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/extension/observer/ecsobserver#metrics and if user care about those errors, they can set alarm on those metrics.

Or we can have a strict mode or development mode to exit on error. It's hard to exit on error as well, it seems there is no good way for extension to tell the service it has crashed because [Start](https://github.com/open-telemetry/opentelemetry-collector/blob/main/component/component.go#L46) starts the extension in background.